File: vulnerable.js
/**
 * @fileoverview Vulnerable JavaScript code samples for SAST testing
 * @author Test Suite
 * @created 2025-01-21
 */

// SQL Injection vulnerability
function getUserData(userId) {
  const query = `SELECT * FROM users WHERE id = ${userId}`;
  return db.query(query);
}

// XSS vulnerability - innerHTML
function displayMessage(message) {
  document.getElementById('message').innerHTML = message;
}

// XSS vulnerability - dangerouslySetInnerHTML (React)
function ReactComponent({ userInput }) {
  return <div dangerouslySetInnerHTML={{ __html: userInput }} />;
}

// Command Injection
function executeCommand(command) {
  const exec = require('child_process').exec;
  exec(command, (error, stdout, stderr) => {
    console.log(stdout);
  });
}

// Code Injection - eval
function processData(data) {
  return eval(data);
}

// Code Injection - Function constructor
function createFunction(code) {
  return new Function(code);
}

// Hardcoded secrets
const apiKey = 'sk_live_51H3ll0W0rld123456789';
const password = 'admin123';
const secret = 'mySecretKey123';

// Path Traversal
function readFile(filename) {
  const fs = require('fs');
  return fs.readFileSync('/var/www/data/' + filename, 'utf8');
}

// Insecure random number generation
function generateToken() {
  return Math.random().toString(36);
}

// Weak crypto - MD5
const crypto = require('crypto');
function hashPassword(password) {
  return crypto.createHash('md5').update(password).digest('hex');
}

// Unsafe deserialization
function deserializeData(data) {
  return JSON.parse(data);
}

// document.write XSS
function writeContent(content) {
  document.write(content);
}

// Unsafe URL redirect
function redirect(url) {
  window.location.href = url;
}

File: vulnerable.rs
/**
 * Vulnerable Rust code samples for SAST testing
 * Author: Test Suite
 * Created: 2025-01-21
 */

use std::process::Command;
use std::fs;
use std::path::PathBuf;
use md5::{Md5, Digest};

// SQL Injection vulnerability (if using string concatenation)
fn get_user_data(user_id: &str) -> String {
    format!("SELECT * FROM users WHERE id = {}", user_id)
}

// Command Injection
fn execute_command(command: &str) -> std::io::Result<()> {
    Command::new("sh")
        .arg("-c")
        .arg(command)
        .output()?;
    Ok(())
}

// Hardcoded secrets
const API_KEY: &str = "sk_live_51H3ll0W0rld123456789";
const PASSWORD: &str = "admin123";
const SECRET: &str = "mySecretKey123";

// Path Traversal
fn read_file(filename: &str) -> std::io::Result<String> {
    let path = PathBuf::from("/var/www/data").join(filename);
    fs::read_to_string(path)
}

// Weak crypto - MD5
fn hash_password(password: &str) -> String {
    let mut hasher = Md5::new();
    hasher.update(password.as_bytes());
    format!("{:x}", hasher.finalize())
}

// Unsafe unwrap (potential panic)
fn get_value() -> i32 {
    let result: Option<i32> = Some(42);
    result.unwrap()  // Could panic if None
}

// Command Injection with Command::new
fn run_command(cmd: &str) -> std::io::Result<()> {
    Command::new("bash")
        .arg("-c")
        .arg(cmd)
        .spawn()?;
    Ok(())
}

